core/src/main/java/com/graphhopper/routing/HeadingResolver.java

```diff
@@ -59,16 +60,46 @@ public IntArrayList getEdgesWithDifferentHeading(int baseNode, double heading) {
         return edges;
     }
 
+    /**
+     * This method returns true if the specified heading is parallel to the specified edgeState (antiparallel isn't tested).
+     * Note that only the road segments near the specified pointNearHeading are checked for parallelism (<20m) and that
+     * a angle difference of 30° is accepted.
+     */
+    public static boolean isHeadingNearlyParallel(EdgeIteratorState edgeState, double heading, GHPoint pointNearHeading) {
+        DistanceCalc calcDist = DistanceCalcEarth.DIST_EARTH;
+        double xAxisAngle = AngleCalc.ANGLE_CALC.convertAzimuth2xaxisAngle(heading);
+        PointList points = edgeState.fetchWayGeometry(FetchMode.ALL);
+        int closestPoint = -1;
+        // TODO should we use the same default like for gpx_accuracy which is 40m?
+        double closestDistance = 20; // skip road segments that are too far away from pointNearHeading
+        double angleDifference = 30;
+        for (int i = 1; i < points.size(); i++) {
+            double fromLat = points.getLat(i - 1), fromLon = points.getLon(i - 1);
+            double toLat = points.getLat(i), toLon = points.getLon(i);
+            double distance = calcDist.validEdgeDistance(pointNearHeading.lat, pointNearHeading.lon, fromLat, fromLon, toLat, toLon)
+                    ? calcDist.calcDenormalizedDist(calcDist.calcNormalizedEdgeDistance(pointNearHeading.lat, pointNearHeading.lon, fromLat, fromLon, toLat, toLon))
+                    : calcDist.calcDist(fromLat, fromLon, pointNearHeading.lat, pointNearHeading.lon);
+
+            if (distance <= closestDistance) {
+                closestDistance = distance;
+                closestPoint = i;
+            }
+        }
+        if (closestPoint >= 0) {
+            double fromLat = points.getLat(closestPoint - 1), fromLon = points.getLon(closestPoint - 1);
+            double toLat = points.getLat(closestPoint), toLon = points.getLon(closestPoint);
+            double orientation = AngleCalc.ANGLE_CALC.calcOrientation(fromLat, fromLon, toLat, toLon);
+            orientation = AngleCalc.ANGLE_CALC.alignOrientation(xAxisAngle, orientation);
+            return Math.abs(orientation - xAxisAngle) < Math.toRadians(angleDifference);
+        }
+        return false;
```

Comment:
Early out?

```suggestion
        if (closestPoint == -1) {
            return false;
        }
        double fromLat = points.getLat(closestPoint - 1), fromLon = points.getLon(closestPoint - 1);
        double toLat = points.getLat(closestPoint), toLon = points.getLon(closestPoint);
        double orientation = AngleCalc.ANGLE_CALC.calcOrientation(fromLat, fromLon, toLat, toLon);
        orientation = AngleCalc.ANGLE_CALC.alignOrientation(xAxisAngle, orientation);
        return Math.abs(orientation - xAxisAngle) < Math.toRadians(angleDifference);
```

Reply:
IMO the `if` is short enough and the `return`s are closer this way.